### PR TITLE
CORE-2237

### PR DESCRIPTION
**Pull request by jeapostrophe**

### CODE REVIEW

File: examples/erc-721/index.rsh [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        }
+      })
+      .api_(I.safeTransferFrom1, (from_, to, tokenId, data) => {
+        check(isApprovedOrOwner(this, tokenId), "ERC721::safeTransferFrom: transfer caller is not owner nor approved");
+        transferChecks(from_, to, tokenId);
+        return [ (k) => {
+          transfer_(from_, to, tokenId);
+          const to_ctc = remote(to, ERC721TokenReceiverI);
+          const mv = to_ctc.onERC721Received(getContract(), from_, tokenId, data);
+          ensure(mv == Bytes.fromHex('0x150b7a02'));
```
Comment (chrisnevers) [inline examples/erc-721/index.rsh:217]: nit: comment/variable explaining that this hex means `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`

**Review by chrisnevers — DISMISSED** [review]
The unimplemented features aside, it looks good to me. The `mixin` seems really useful. Maybe comments about the meaning of hex IDs would be useful

**Comment by willghatch** [comment]
For future readers, note that some comments about this are in https://github.com/reach-sh/reach-lang/pull/1392 because I was silly and opened another PR.

File: examples/ERC721/index.rsh [hunk 2]
```diff
@@ -1,2 +1,307 @@
...
+          owners[tokenId] = to;
+          E.Transfer(from_, to, tokenId);
+        }
+      })
+      .api_(I.safeTransferFrom1, (from_, to, tokenId, data) => {
+        check(isApprovedOrOwner(this, tokenId), "ERC721::safeTransferFrom: transfer caller is not owner nor approved");
+        transferChecks(from_, to, tokenId);
+        return [ (k) => {
+          transfer_(from_, to, tokenId);
+          // TODO - this should be a remote call if and only if the to address is a contract.  We need to use the future feature Contract.fromAddress here to determine that.
```
Comment (jeapostrophe) [inline examples/ERC721/index.rsh:220]: I believe that this function only makes sense when `to` is a Contract, so we should just change the type, and then maybe it is not `Contract.fromAddress` that we want but `Address.fromContract` that we want

File: examples/ERC721/index.rsh [hunk 3]
```diff
@@ -1,2 +1,307 @@
...
+          //const to_ctc = remote(to, ERC721TokenReceiverI);
+          //const mv = to_ctc.onERC721Received(getContract(), from_, tokenId, data);
+          //// This hex string is bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))
+          //ensure(mv == Bytes_fromHex('0x150b7a02'));
+          k(null);
+          return [ ];
+        }];
+      })
+      .api_(I.safeTransferFrom2, (from_, to, tokenId) => {
+        // TODO - this check is important, but no transfers are working with it on.  So something is wrong with it (or with how I'm calling this method).
```
Comment (jeapostrophe) [inline examples/ERC721/index.rsh:230]: What error do you get?

File: examples/ERC721/index.rsh [hunk 4]
```diff
@@ -1,2 +1,307 @@
...
+          operatorApprovals[[owner, operator]] = approved;
+          E.ApprovalForAll(owner, operator, approved);
+          k(null);
+          return [ ];
+        }];
+      })
+      .api_(I.mint, (to, tokenId) => {
+        check(to != zeroAddr, "Cannot mint to zero address");
+        check(!tokenExists(tokenId), "Token already exists");
+        // TODO - shouldn't minting only be available to the deployer or something like that?
```
Comment (jeapostrophe) [inline examples/ERC721/index.rsh:284]: My guess is that the spec doesn't mandate it that way, but you should add that

File: examples/ERC721/index.rsh [hunk 5]
```diff
@@ -1,2 +1,307 @@
...
+        return [ (k) => {
+          balances[to] = fromMaybe(balances[to], () => 1, (x) => x + 1);
+          owners[tokenId] = to;
+          E.Transfer(zeroAddr, to, tokenId);
+          k(null);
+          return [ ];
+        }];
+      })
+      .api_(I.burn, (tokenId) => {
+        // TODO - I don't see any check here that only the owner can burn it...
```
Comment (jeapostrophe) [inline examples/ERC721/index.rsh:294]: Ditto

File: examples/ERC721/jay-small-test.mjs [hunk 6]
```diff
@@ -0,0 +1,110 @@
+import { loadStdlib, util } from '@reach-sh/stdlib';
```
Comment (jeapostrophe) [inline examples/ERC721/jay-small-test.mjs:1]: This file was Nevers' original test. I think it should be deleted and instead we should only test with the general 721 test suite

File: examples/ERC721/index.rsh [hunk 7]
```diff
@@ -1,2 +1,307 @@
...
+
+        const modAt = (dict, key, f) => {
+          const value = dict[key];
+          fromMaybe(value, () => {}, (x) => { dict[key] = f(x); });
+        }
+
+        V.balanceOf.set((owner) => {
+          check(owner != zeroAddr, "ERC721::balanceOf: Address zero is not a valid owner");
+          const m_bal = balances[owner];
+          // TODO - I don't understand why we would want this check.  I think it's wrong, we want to return 0 if they own none.
```
Comment (jeapostrophe) [inline examples/ERC721/index.rsh:155]: My guess is that the spec says this

File: examples/ERC721/index.rsh [hunk 8]
```diff
@@ -1,2 +1,307 @@
...
+  }],
+});
+
+const ERC721Enumerable = mixin({
+  Base: ERC721,
+  IDs: [ Bytes_fromHex('0x780e9d63'), ],
+  View: [{
+    totalSupply: UInt,
+    tokenByIndex: Fun([UInt], UInt),
+    tokenOfOwnerByIndex: Fun([Address, UInt], UInt),
```
Comment (jeapostrophe) [inline examples/ERC721/index.rsh:82]: Can you add these functions?

File: examples/ERC721/index.rsh (hunk 3 above)
Comment (willghatch) [reply examples/ERC721/index.rsh:230]:
```
/stdlib/node_modules/@ethersproject/logger/lib/index.js:233
        var error = new Error(message);
                    ^

Error: transaction failed [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (transactionHash="0xc611596f18bf7ea98a9b54e035f7552773f71366ea44468242514a607b31a494", transaction={"type":2,"chainId":1337,"nonce":34,"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"maxFeePerGas":{"type":"BigNumber","hex":"0x59682f10"},"gasPrice":null,"gasLimit":{"type":"BigNumber","hex":"0x4c4b40"},"to":"0x32a33D22eB5c4cAa9228E2d5BF5509bFd6f38250","value":{"type":"BigNumber","hex":"0x00"},"data":"0x42842e0e000000000000000000000000cf3e34cbf98729c78c0310d0745c83efe0b01bd80000000000000000000000003ffd5f6e6be1589aad16d4d3d5c15fab58dea46a0000000000000000000000000000000000000000000000000000000000000001","accessList":[],"hash":"0xc611596f18bf7ea98a9b54e035f7552773f71366ea44468242514a607b31a494","v":0,"r":"0x987a11b4aba1da03ef81e416346b8b4bc3584e169f4c3df43d9c929e54cd69f9","s":"0x039f6cbaf1e3d8bf5a0ce7839d7d777236ffaabaaa3a4c6854bc4fd5dd1e4968","from":"0xCf3E34cBf98729C78c0310d0745c83EfE0b01bd8","confirmations":0}, receipt={"to":"0x32a33D22eB5c4cAa9228E2d5BF5509bFd6f38250","from":"0xCf3E34cBf98729C78c0310d0745c83EfE0b01bd8","contractAddress":null,"transactionIndex":0,"gasUsed":{"type":"BigNumber","hex":"0x014054"},"logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","blockHash":"0x529c5df7ca904358cc68efee2f7343b6373ab1385c6c0c590f2448dbf2dffe4b","transactionHash":"0xc611596f18bf7ea98a9b54e035f7552773f71366ea44468242514a607b31a494","logs":[],"blockNumber":3326,"confirmations":1,"cumulativeGasUsed":{"type":"BigNumber","hex":"0x014054"},"effectiveGasPrice":{"type":"BigNumber","hex":"0x59682f08"},"status":0,"type":2,"byzantium":true}, code=CALL_EXCEPTION, version=providers/5.6.8)
    at Logger.makeError (/stdlib/node_modules/@ethersproject/logger/lib/index.js:233:21)
    at Logger.throwError (/stdlib/node_modules/@ethersproject/logger/lib/index.js:242:20)
    at JsonRpcProvider.<anonymous> (/stdlib/node_modules/@ethersproject/providers/lib/base-provider.js:1683:36)
    at step (/stdlib/node_modules/@ethersproject/providers/lib/base-provider.js:48:23)
    at Object.next (/stdlib/node_modules/@ethersproject/providers/lib/base-provider.js:29:53)
    at fulfilled (/stdlib/node_modules/@ethersproject/providers/lib/base-provider.js:20:58)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  reason: 'transaction failed',
  code: 'CALL_EXCEPTION',
  transactionHash: '0xc611596f18bf7ea98a9b54e035f7552773f71366ea44468242514a607b31a494',
  transaction: {
    type: 2,
    chainId: 1337,
    nonce: 34,
    maxPriorityFeePerGas: BigNumber { _hex: '0x59682f00', _isBigNumber: true },
    maxFeePerGas: BigNumber { _hex: '0x59682f10', _isBigNumber: true },
    gasPrice: null,
    gasLimit: BigNumber { _hex: '0x4c4b40', _isBigNumber: true },
    to: '0x32a33D22eB5c4cAa9228E2d5BF5509bFd6f38250',
    value: BigNumber { _hex: '0x00', _isBigNumber: true },
    data: '0x42842e0e000000000000000000000000cf3e34cbf98729c78c0310d0745c83efe0b01bd80000000000000000000000003ffd5f6e6be1589aad16d4d3d5c15fab58dea46a0000000000000000000000000000000000000000000000000000000000000001',
    accessList: [],
    hash: '0xc611596f18bf7ea98a9b54e035f7552773f71366ea44468242514a607b31a494',
    v: 0,
    r: '0x987a11b4aba1da03ef81e416346b8b4bc3584e169f4c3df43d9c929e54cd69f9',
    s: '0x039f6cbaf1e3d8bf5a0ce7839d7d777236ffaabaaa3a4c6854bc4fd5dd1e4968',
    from: '0xCf3E34cBf98729C78c0310d0745c83EfE0b01bd8',
    confirmations: 0,
    wait: [Function (anonymous)]
  },
  receipt: {
    to: '0x32a33D22eB5c4cAa9228E2d5BF5509bFd6f38250',
    from: '0xCf3E34cBf98729C78c0310d0745c83EfE0b01bd8',
    contractAddress: null,
    transactionIndex: 0,
    gasUsed: BigNumber { _hex: '0x014054', _isBigNumber: true },
    logsBloom: '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    blockHash: '0x529c5df7ca904358cc68efee2f7343b6373ab1385c6c0c590f2448dbf2dffe4b',
    transactionHash: '0xc611596f18bf7ea98a9b54e035f7552773f71366ea44468242514a607b31a494',
    logs: [],
    blockNumber: 3326,
    confirmations: 1,
    cumulativeGasUsed: BigNumber { _hex: '0x014054', _isBigNumber: true },
    effectiveGasPrice: BigNumber { _hex: '0x59682f08', _isBigNumber: true },
    status: 0,
    type: 2,
    byzantium: true
  }
}
ERROR: 1
```

File: examples/ERC721/index.rsh (hunk 8 above)
Comment (willghatch) [reply examples/ERC721/index.rsh:82]: Will do.

**Comment by willghatch** [comment]
I've found the reason  why it's not working when I check for `isApprovedOrOwner` -- the value of `this` ends up being the address of the contract, not the caller.  I'm confused about that.  How would that happen?

**Comment by willghatch** [comment]
I did look at it, and it uses `msg.sender`, and the docs for that seem like it's the right thing.  In the client code, the JS contract object has `ctc.address` of the contract, but it also has (I forget in what field exactly) the correct user address as signer, so I would think it would be the address that the contract would see as the sender.  Anyway, I'll keep looking at it today, but I'm stumped so far.

**Comment by willghatch** [comment]
So as this PR stands, the contract passes the test suite with 2 caveats:

• The ERC721Enumerable section is still disabled (I said I would implement it in the contract but I haven't yet).
• The test suite doesn't yet test `burn`, `approve`, `setApprovalForAll`, or anything related to using approved users instead of owners, or `safeTransferFrom` to a contract (and thus actually making the remote call).

**Review by jeapostrophe — COMMENTED** [review]
I think this should be merged ASAP and then the extra features added after

**Comment by willghatch** [comment]
With that golden file pushed, it should hopefully be ready.  Do you want me to do those things next, or do my other sprint tasks first?

**Pull request merged**